src/iterators/chunked_read_ahead_iterator.rs

```diff
@@ -0,0 +1,393 @@
+    /// underlying buffer elements.
+    ///
+    /// # Panics
+    ///
+    /// - panics if the underlying iterator panics
+    /// - panics if sending final chunk after an internal panic fails
+    /// - panics if sending the None kill signal to the recieiver after an internal panic fails
+    /// - panics if the spawned thread fails to spawn
+    pub fn new<I>(mut inner: I, chunk_size: usize, num_chunk_buffer_size: usize) -> Self
+    where
+        I: Iterator<Item = T> + Send + 'static,
+    {
+        assert_ne!(chunk_size, 0, "Chunk size cannot be zero!");
+        assert_ne!(num_chunk_buffer_size, 0, "Number of buffered chunks cannot be zero!");
+
+        // Create a channel over which we can send our chunks of ``T``
+        let (sender, receiver) = sync_channel(num_chunk_buffer_size);
+
+        // Create our spawned thread, holding on to the resulting handle for downstream error
+        // management.
+        let join_handle: JoinHandle<()> = thread::Builder::new()
+            .name("chunked_read_ahead_thread".to_owned())
+            .spawn(move || {
+                loop {
+                    let mut chunk = Vec::with_capacity(chunk_size);
+                    for _ in 0..chunk_size {
+                        match catch_unwind(AssertUnwindSafe(|| inner.by_ref().next())) {
+                            Ok(Some(val)) => chunk.push(val),
+                            Ok(None) => break,
+                            Err(e) => {
+                                sender
+                                    .send(Some(chunk))
+                                    .expect("Error sending final chunk before internal panic");
+                                sender
+                                    .send(None)
+                                    .expect("Error sending final None before internal panic");
+                                panic!("{}", extract_info_from_catch_unwind_panic(e));
+                            }
+                        }
+                    }
+                    if chunk.is_empty() {
+                        break;
+                    }
+                    let send_result = sender.send(Some(chunk));
+                    if send_result.is_err() {
+                        // TODO - Logging?
+                        break;
+                    }
+                }
+                // TODO - Logging?
+                let _send_result = sender.send(None);
+            })
+            .expect("failed to spawn chunked read ahead thread");
+
+        // Store the necessary objects on ``Self``
+        Self {
+            current_chunk: Vec::new().into_iter(),
+            receiver: Some(receiver),
+            join_handle: Some(join_handle),
+        }
+    }
+}
+
+impl<T> Drop for ChunkedReadAheadIterator<T>
+where
+    T: Send + 'static,
+{
+    fn drop(&mut self) {
+        // Make sure our reciever is dropped so our spawned thread shuts down.
+        self.receiver = None;
```

Comment:
I feel like the following is clearer in intent
```suggestion
        drop(self.receiver.take());
```

Reply:
And actually, I think the whole custom `Drop` impl can go away now, and as part of that the Receiver can be stored without an Option, right?